pkg/lexer/lexer.go

```diff
@@ -1,5 +1,215 @@
+
+func (l *Lexer) peekRune() rune {
+	if l.pos+1 >= len(l.input) {
+		return 0
+	}
+	return l.input[l.pos+1]
+}
+
+func (l *Lexer) Next() *Token {
+	l.readRune()
+	l.skipWhitespace()
+
+	tok := &Token{
+		Offset: l.pos,
+		Line:   l.line,
+		Col:    l.col,
+	}
+	switch l.r {
+	case '=':
+		if l.peekRune() == '=' {
+			l.readRune()
+			return tok.SetType(EQ)
+		}
+		return tok.SetType(ASSIGN)
+	case '+':
+		return tok.SetType(PLUS)
+	case '-':
+		return tok.SetType(MINUS)
+	case '!':
+		if l.peekRune() == '=' {
+			l.readRune()
+			return tok.SetType(NOT_EQ)
+		}
+		return tok.SetType(BANG)
+	case '/':
+		if l.peekRune() == '/' {
+			return tok.SetType(COMMENT).SetLiteral(l.readComment())
+		}
+		return tok.SetType(SLASH)
+	case '*':
+		return tok.SetType(ASTERISK)
+	case '<':
+		if l.peekRune() == '=' {
+			l.readRune()
+			return tok.SetType(LTEQ)
+		}
+		return tok.SetType(LT)
+	case '>':
+		if l.peekRune() == '=' {
+			l.readRune()
+			return tok.SetType(GTEQ)
+		}
+		return tok.SetType(GT)
+	case ':':
+		if l.peekRune() == '=' {
+			l.readRune()
+			return tok.SetType(DECLARE)
+		}
+		return tok.SetType(COLON)
+	case '{':
+		return tok.SetType(LCURLY)
+	case '}':
+		return tok.SetType(RCURLY)
+	case '(':
+		return tok.SetType(LPAREN)
+	case ')':
+		return tok.SetType(RPAREN)
+	case '[':
+		return tok.SetType(LBRACKET)
+	case ']':
+		return tok.SetType(RBRACKET)
+	case '\n':
+		return tok.SetType(NL)
+	case '.':
+		if l.peekRune() == '.' && l.peekRune2() == '.' {
+			l.readRune()
+			l.readRune()
+			return tok.SetType(DOT3)
+		}
+		return tok.SetType(DOT)
+	case '"':
+		literal := l.readString()
+		// EOF or NL before closing `"`
+		if l.r != '"' {
+			return tok.SetType(ILLEGAL).SetLiteral(`"`)
+		}
+		return tok.SetType(STRING_LIT).SetLiteral(literal)
+	case 0:
+		return tok.SetType(EOF)
+	}
+	if isLetter(l.r) {
+		lit := l.readIdent()
+		tokenType := LookupKeyword(lit)
+		if tokenType == IDENT {
+			return tok.SetType(IDENT).SetLiteral(lit)
+		}
+		return tok.SetType(tokenType)
+	}
+	if isDigit(l.r) {
+		return tok.SetType(NUM_LIT).SetLiteral(l.readNum())
+	}
+
+	return tok.SetType(ILLEGAL).SetLiteral(string(l.r))
+}
+
+func (l *Lexer) peekRune2() rune {
+	if l.pos+2 >= len(l.input) {
+		return 0
+	}
+	return l.input[l.pos+2]
+}
+
+func (l *Lexer) readComment() string {
+	pos := l.pos
+	pr := l.peekRune()
+	for pr != 0 && pr != '\n' {
+		l.readRune()
+		pr = l.peekRune()
+	}
+	return string(l.input[pos : l.pos+1])
+}
+
+func (l *Lexer) readString() string {
+	pos := l.pos + 1
+	for {
+		l.readRune()
+		pr := l.peekRune()
+		if pr == '"' && l.r != '\\' {
+			l.readRune()
+			break
+		}
+		if pr == 0 || pr == '\n' {
+			break
+		}
+	}
+	s := string(l.input[pos:l.pos])
+	return strings.ReplaceAll(s, `\"`, `"`)
+}
+
+func (l *Lexer) readNum() string {
+	pos := l.pos
+	pr := l.peekRune()
+	for isDigit(pr) || pr == '.' {
+		l.readRune()
+		pr = l.peekRune()
+	}
+	return string(l.input[pos : l.pos+1])
+}
+
+func (l *Lexer) readIdent() string {
```

Comment:
`readComment`, `readNum` and `readIdent` could be done as (just an idea):
```
func (l *Lexer) readWhile(pred func(rune) bool) string {
        pos := l.pos
        for pr := l.peekRune(); pred(pr); pr = l.peekRune() {
                l.readRune()
        }
        return string(l.input[pos : l.pos+1])
}

func (l *Lexer) readComment() string {
        return l.readWhile(func(r rune) bool { return r != 0 && r != '\n' })
}

func (l *Lexer) readNum() string {
        return l.readWhile(func(r rune) bool { return isDigit(r) || r == '.' })
}

func (l *Lexer) readIdent() string {
        return l.readWhile(func(r rune) bool { return isLetter(r) || unicode.IsDigit(r) })
}
```

It's a shame anonymous Go functions have so much boilerplate. Would be nicer with an arrow syntax.

Reply:
That's super nice, i like that a lot!

Reply:
🏆

Reply:
Yeah, I tried readString but it doesn't fit the mould because it needs to look ahead.

Reply:
ok, thanks - found the same.